src/main/java/fi/vm/sade/varda/rekisterointi/configuration/HakijaWebSecurityConfig.java

```diff
@@ -109,39 +118,50 @@ public AuthenticationProvider hakijaAuthenticationProvider() {
+    private static class HakijaAuthenticationFilter extends AbstractAuthenticationProcessingFilter {
 
-        public ShibbolethAuthenticationFilter(String defaultFilterProcessesUrl) {
+        private final TicketValidator oppijaticketValidator;
+        private final OphProperties properties;
+
+
+        public HakijaAuthenticationFilter(String defaultFilterProcessesUrl, TicketValidator oppijaticketValidator, OphProperties properties) {
             super(defaultFilterProcessesUrl);
+            this.properties = properties;
+            this.oppijaticketValidator = oppijaticketValidator;
         }
 
         @Override
         public Authentication attemptAuthentication(HttpServletRequest request, HttpServletResponse response) throws AuthenticationException, IOException {
-            String nationalIdentificationNumber = Optional.ofNullable(request.getHeader("nationalidentificationnumber"))
-                    .or(() -> Optional.ofNullable(request.getParameter("hetu"))) // for easier development
-                    .orElseThrow(() -> new PreAuthenticatedCredentialsNotFoundException("Unable to authenticate because required header doesn't exist"));
-
-            String firstName = Optional.ofNullable(request.getHeader("firstname"))
-                    .map(str -> new String(str.getBytes(StandardCharsets.ISO_8859_1), StandardCharsets.UTF_8))
+            String ticket = Optional.ofNullable(request.getParameter("ticket"))
+                    .orElseThrow(() -> new PreAuthenticatedCredentialsNotFoundException("Unable to authenticate because required param doesn't exist"));
+            Map<String, Object> casPrincipalAttributes = null;
+            try { //Validate uses Cas:s serviceValidate endpoint and parses the result xml into custom Attributes.
+                String hakijaLoginurl = properties.url("varda-rekisterointi.hakija.login");
+                casPrincipalAttributes = oppijaticketValidator.validate(ticket, hakijaLoginurl).getPrincipal().getAttributes();
+            } catch (TicketValidationException e) {
+                throw new AuthenticationCredentialsNotFoundException("Unable to authenticate because required param doesn't exist");
+            }
+            String nationalIdentificationNumber = Optional.ofNullable((String)casPrincipalAttributes.get("nationalIdentificationNumber"))
+                    .orElseThrow(() -> new PreAuthenticatedCredentialsNotFoundException("Unable to authenticate because required param doesn't exist"));
+            String surname = Optional.ofNullable((String)casPrincipalAttributes.get("sn"))
                     .orElse("");
-            String surname = Optional.ofNullable(request.getHeader("sn"))
-                    .map(str -> new String(str.getBytes(StandardCharsets.ISO_8859_1), StandardCharsets.UTF_8))
+            String firstName = Optional.ofNullable((String)casPrincipalAttributes.get("firstName"))
```

Comment:
There is quite lots going on in this method. It would be rather easy to break down in smaller functional chunks with meaningful names. This would also make it easier to write unit tests (where are those btw?)
This was really quickly done with "refactor -> extract method" functionality of my IDE:
      
```
       @Override
        public Authentication attemptAuthentication(HttpServletRequest request, HttpServletResponse response) throws AuthenticationException, IOException {
            String ticket = resolveTicket(request);
            try {
                String hakijaLoginurl = properties.url("varda-rekisterointi.hakija.login");
                Map<String, Object> casPrincipalAttributes = oppijaticketValidator.validate(ticket, hakijaLoginurl).getPrincipal().getAttributes();
                return getAuthenticationManager().authenticate(createrAuthRequest(request, casPrincipalAttributes));
            } catch (TicketValidationException e) {
                throw new AuthenticationCredentialsNotFoundException("Unable to authenticate because required param doesn't exist");
            }
        }

        private PreAuthenticatedAuthenticationToken createrAuthRequest(HttpServletRequest request, Map<String, Object> casPrincipalAttributes) {
            String nationalIdentificationNumber = Optional.ofNullable((String) casPrincipalAttributes.get("nationalIdentificationNumber"))
                    .orElseThrow(() -> new PreAuthenticatedCredentialsNotFoundException("Unable to authenticate because required param doesn't exist"));
            String surname = Optional.ofNullable((String) casPrincipalAttributes.get("sn"))
                    .orElse("");
            String firstName = Optional.ofNullable((String) casPrincipalAttributes.get("firstName"))
                    .orElse("");

            PreAuthenticatedAuthenticationToken authRequest = new PreAuthenticatedAuthenticationToken(nationalIdentificationNumber, "N/A");
            List<? extends GrantedAuthority> authorities = singletonList(new SimpleGrantedAuthority(String.format("ROLE_%s", HAKIJA_ROLE)));
            authRequest.setDetails(new CasOppijaAuthenticationDetails(request, authorities, firstName, surname));
            return authRequest;
        }

        private String resolveTicket(HttpServletRequest request) {
            return Optional.ofNullable(request.getParameter("ticket"))
                    .orElseThrow(() -> new PreAuthenticatedCredentialsNotFoundException("Unable to authenticate because required param doesn't exist"));
        }
```